File: Classes/PSBookmarks.m
//
//  PSBookmarks.m
//  PocketSword
//
//  Created by Nic Carter on 12/01/11.
//  Copyright 2011 CrossWire Bible Society. All rights reserved.
//

#import "PSBookmarks.h"
#import "globals.h"
#import "PSBookmark.h"
#import "PSBookmarkFolder.h"
#import "PSModuleController.h"

@implementation PSBookmarks

static PSBookmarks *psDefaultBookmarks;// = nil;
/** the singleton instance */
+ (PSBookmarks *)defaultBookmarks {
	//DLog(@"\nattempting access to our bookmarks...");
    if(psDefaultBookmarks == nil) {
		//DLog(@"\nOur bookmarks object doesn't exist! :P");
        psDefaultBookmarks = [[PSBookmarks alloc] initLocalBookmarks];
    }
	return psDefaultBookmarks;
}

//+ (void)initialize {
//	if (self == [PSBookmarks class]) {
//		DLog(@"\ninitialize");
//        psDefaultBookmarks = [[self alloc] initLocalBookmarks];
//    }
//}

+ (BOOL)addBookmarkObject:(PSBookmarkObject*)bookmark withFolderString:(NSString*)folderString {
	BOOL ret = NO;
	PSBookmarks *bookmarks = [PSBookmarks defaultBookmarks];
	if(!folderString || [folderString isEqualToString:@""]) {
		[bookmarks addChild:bookmark];
		ret = YES;
	} else {
		PSBookmarkFolder *parentFolder = [PSBookmarks getBookmarkFolderForFolderString:folderString];
		[parentFolder addChild:bookmark];
		ret = YES;
	}
	[PSBookmarks saveBookmarksToFile];
	return ret;
}

+ (BOOL)addBookmarkWithRef:(NSString*)r name:(NSString*)n folderString:(NSString*)folderString {
	NSDate *date = [NSDate date];
	PSBookmark *bookmark = [[PSBookmark alloc] initWithName:n dateAdded:date dateLastAccessed:date bibleReference:r];
	BOOL ret = [PSBookmarks addBookmarkObject:bookmark withFolderString:folderString];
	[bookmark release];
	return ret;
}

+ (void)deleteBookmark:(NSString*)n fromFolderString:(NSString*)folderString {
	PSBookmarkFolder *parent = [PSBookmarks getBookmarkFolderForFolderString:folderString];
	NSMutableArray *array = [parent.children mutableCopy];
	// need to identify which is the correct child
	for(PSBookmarkObject *obj in array) {
		if([obj.name isEqualToString:n]) {
			[array removeObject:obj];
			break;
		}
	}
	parent.children = array;
	[array release];
	[PSBookmarks saveBookmarksToFile];
}

+ (PSBookmarkFolder*)getBookmarkFolderForFolderString:(NSString*)folderString {
	PSBookmarks *bookmarks = [PSBookmarks defaultBookmarks];
	if(!folderString || [folderString isEqualToString:@""] || [folderString isEqualToString:NSLocalizedString(@"BookmarksTitle", @"")]) {
		return bookmarks;
	}
	NSArray *folders = [folderString componentsSeparatedByString:PSFolderSeparatorString];
	PSBookmarkFolder *parentFolder = bookmarks;
	for(int folderCount=0; folderCount<[folders count];folderCount++) {
		BOOL foundFolder = NO;
		for(int kidNumber=0;kidNumber<[parentFolder.children count];kidNumber++) {
			if([((PSBookmarkObject*)[parentFolder.children objectAtIndex:kidNumber]).name isEqualToString:[folders objectAtIndex:folderCount]]) {
				parentFolder = [parentFolder.children objectAtIndex:kidNumber];
				foundFolder = YES;
				break;
			}
		}
		if(!foundFolder) {
			// if we don't find the next folder, just place the bookmark here.  This shouldn't be possible!
			ALog(@"getBookmarkFolderForFolderString: couldn't find the folder - %@", [folders objectAtIndex:folderCount]);
			break;
		}
	}
	return parentFolder;
}

+ (NSString *)getHighlightRGBColourStringForBookAndChapterRef:(NSString*)bookAndChapterRef withVerse:(NSInteger)verse {
	return [[PSBookmarks defaultBookmarks] getHighlightRGBColourStringForBookAndChapterRef:bookAndChapterRef withVerse:[NSString stringWithFormat:@"%d", verse]];
}

+ (NSMutableArray *)getBookmarksForCurrentRef {
	NSString *currentRef = [PSModuleController createRefString:[PSModuleController getCurrentBibleRef]];
	return [PSBookmarks getBookmarksForBookAndChapterRef:currentRef];
}

+ (NSMutableArray *)getBookmarksForBookAndChapterRef:(NSString*)bookAndChapterRef {
	PSBookmarks *bookmarks = [PSBookmarks defaultBookmarks];
	return [bookmarks getBookmarksForBookAndChapterRef:bookAndChapterRef];
}

- (PSBookmarkObject *)parseArray:(NSArray *)array {
	if(!array)
		return nil;
	else if([array count] == 0)
		return nil;
	
	NSString *n = [array objectAtIndex:0];
	NSDate *da = [array objectAtIndex:1];
	NSDate *dla = [array objectAtIndex:2];
	NSString *folderString = [array objectAtIndex:3];
	if([folderString boolValue]) {
		//tis a folder
		NSString *rgb = [array objectAtIndex:4];
		if([rgb isEqualToString:@""]) {
			rgb = nil;
		}
		NSArray *kids = [array objectAtIndex:5];
		NSMutableArray *kidsArray = [NSMutableArray arrayWithCapacity:[kids count]];
		for(NSArray *child in kids) {
			PSBookmarkObject *kid = [self parseArray:child];
			[kidsArray addObject:kid];
			[kid release];
		}
		return [[PSBookmarkFolder alloc] initWithName:n dateAdded:da dateLastAccessed:dla rgbHexString:rgb children:kidsArray];
	} else {
		//tis a bookmark
		NSString *r = [array objectAtIndex:4];
		return [[PSBookmark alloc] initWithName:n dateAdded:da dateLastAccessed:dla bibleReference:r];
	}
}

- (void)loadBookmarksFromArray:(NSArray *)dataArray {
	//DLog(@"\nBookmarks: loadBookmarksFromFile");
	NSMutableArray *kidsArray = [NSMutableArray arrayWithCapacity:2];
	if(dataArray) {
		for(NSArray *child in dataArray) {
			PSBookmarkObject *kid = [self parseArray:child];
			[kidsArray addObject:kid];
			[kid release];
		}
        self.children = kidsArray;
    } else {
        self.children = [NSArray array];
    }
	//DLog(@"\n-- Bookmarks: finished loadBookmarksFromFile");
}

- (id)initLocalBookmarks {
	self = [super initWithName:nil dateAdded:nil dateLastAccessed:nil rgbHexString:nil children:nil];
	if(self) {
		self.name = NSLocalizedString(@"BookmarksTitle", @"");
		NSString *bookmarksPath = [DEFAULT_BOOKMARKS_PATH stringByAppendingPathComponent:@"PSBookmarks.plist"];
		NSArray *dataArray = [NSArray arrayWithContentsOfFile:bookmarksPath];
		[self loadBookmarksFromArray:dataArray];
	}
//	DLog(@"\n***\ncreated a local PSBookmarks object...\n***\n");
	return self;
}

- (id)initCloudBookmarks {
	self = [super initWithName:nil dateAdded:nil dateLastAccessed:nil rgbHexString:nil children:nil];
	if(self) {
		self.name = NSLocalizedString(@"BookmarksTitle", @"");
		NSData *data = [@"plistStringToCreateADataThingo" dataUsingEncoding:NSUTF8StringEncoding];
		//format should be NSPropertyListXMLFormat_v1_0
		NSArray *array = [NSPropertyListSerialization
						  propertyListWithData:data
						  options:NSPropertyListImmutable
						  format:NULL
						  error:NULL];
		[self loadBookmarksFromArray:array];
	}
//	DLog(@"\n***\ncreated a cloud PSBookmarks object...\n***\n");
	return self;

	// to write back, use:
//	[NSPropertyListSerialization dataWithPropertyList:array format:NSPropertyListXMLFormat_v1_0 options:NSPropertyListImmutable error:NULL];
//	+ (NSData *)dataWithPropertyList:(id)plist format:(NSPropertyListFormat)format options:(NSPropertyListWriteOptions)opt error:(NSError **)error


}

+ (NSDate *)lastModified:(PSBookmarkObject *)bookmarkObject {
	
	NSDate *returnDate = bookmarkObject.dateLastAccessed;
	if(bookmarkObject.folder) {
		for(PSBookmarkObject *child in ((PSBookmarkFolder *)bookmarkObject).children) {
			NSDate *childDate = [PSBookmarks lastModified:child];
			if([childDate compare:returnDate] == NSOrderedDescending) {
				returnDate = child.dateLastAccessed;
			}
		}
	}
	return returnDate;
	
}

+ (NSDate *)lastModified {
	PSBookmarks *bookmarks = [PSBookmarks defaultBookmarks];
	return [PSBookmarks lastModified:bookmarks];
}

// if our current bookmarks have been modified more recently than the new one we're sent via the cloud
//    combine the bookmarks rather than deleting any.
//    and create a copy of iCloud + local as backups.

+ (NSArray *)parseBookmarkObject:(PSBookmarkObject*)bookmarkObject {
	if(!bookmarkObject)
		return nil;
	int capacity = (bookmarkObject.folder) ? 7 : 5;
	NSMutableArray *ret = [NSMutableArray arrayWithCapacity:capacity];
	[ret addObject:bookmarkObject.name];
	[ret addObject:bookmarkObject.dateAdded];
	[ret addObject:bookmarkObject.dateLastAccessed];
	if(bookmarkObject.folder) {
		[ret addObject:@"YES"];
		if(((PSBookmarkFolder*)bookmarkObject).rgbHexString) {
			[ret addObject:((PSBookmarkFolder*)bookmarkObject).rgbHexString];
		} else {
			[ret addObject:@""];
		}
		NSMutableArray *kids = [NSMutableArray arrayWithCapacity:[((PSBookmarkFolder*)bookmarkObject).children count]];
		for(PSBookmarkObject *child in ((PSBookmarkFolder*)bookmarkObject).children) {
			NSArray *kid = [PSBookmarks parseBookmarkObject:child];
			[kids addObject:kid];
			//[kid release]; -- they're autorelease objects :P
		}
		[ret addObject:kids];
		
	} else {
		[ret addObject:@"NO"];
		[ret addObject:((PSBookmark*)bookmarkObject).ref];
	}
	
	return ret;
}

+ (BOOL)saveBookmarksToFile {
	PSBookmarks *bookmarks = [PSBookmarks defaultBookmarks];
	BOOL ret = NO;
	//DLog(@"\nBookmarks: saveBookmarksToFile");
    NSString *bookmarksPath = [DEFAULT_BOOKMARKS_PATH stringByAppendingPathComponent:@"PSBookmarks.plist"];
	if(bookmarks.children && [bookmarks.children count] > 0) {
		NSMutableArray *data = [NSMutableArray arrayWithCapacity:[bookmarks.children count]];
		for(PSBookmarkObject *child in bookmarks.children) {
			NSArray *kid = [PSBookmarks parseBookmarkObject:child];
			[data addObject:kid];
			//[kid release]; -- they're autorelease objects :P
		}
		ret = [data writeToFile:bookmarksPath atomically:YES];
	} else {
		// no bookmarks left! Write out the file to zeros...
		NSMutableArray *data = [NSMutableArray arrayWithCapacity:1];
		ret = [data writeToFile:bookmarksPath atomically:NO];
	}
	
	DLog(@"\n-- Bookmarks: finished saveBookmarksToFile");
	return ret;
}

+ (void)importBookmarksFromV2 {
	NSArray *oldBookmarks = [[NSUserDefaults standardUserDefaults] arrayForKey: @"bookmarks2"];
	
	if(oldBookmarks) {
		PSBookmarks *bookmarks = [PSBookmarks defaultBookmarks];
		BOOL createImportedFolder = YES;
		for(PSBookmarkObject* obj in bookmarks.children) {
			if([obj.name isEqualToString:NSLocalizedString(@"BookmarksImportedFolderName", @"")]) {
				// if there already exists an @"imported" folder, don't recreate it!
				createImportedFolder = NO;
				break;
			}
		}
		if(createImportedFolder) {
			PSBookmarkFolder *importFolder = [[PSBookmarkFolder alloc] initWithName:NSLocalizedString(@"BookmarksImportedFolderName", @"") dateAdded:[NSDate date] dateLastAccessed:[NSDate date] rgbHexString:nil children:nil];
			[PSBookmarks addBookmarkObject:importFolder withFolderString:nil];
			[importFolder release];
		}
		for(NSString *ref in oldBookmarks) {
			[PSBookmarks addBookmarkWithRef:[PSModuleController createRefString:ref] name:ref folderString:NSLocalizedString(@"BookmarksImportedFolderName", @"")];
		}
		[[NSUserDefaults standardUserDefaults] removeObjectForKey:@"bookmarks2"];
		[[NSUserDefaults standardUserDefaults] synchronize];
	}
}

@end

File: Classes/PSBookmarks.swift
//
//  PSBookmarks.swift
//  PocketSword
//
//  Swift migration step 1.1 — bookmark inheritance chain (ATOMIC, §2A Rule 3).
//  Port of PSBookmarks.{h,mm} (was Obj-C++ only because it imported
//  PSModuleController.h for two thin Foundation seams; those are now extracted
//  to PSRefHelper in AppConstants.swift, so this file carries no C++).
//
//  CRITICAL — the on-disk PSBookmarks.plist positional-array schema is preserved
//  BYTE-FOR-BYTE (locked by PocketSwordTests / PersistedFormatTests):
//    WRITE  +parseBookmarkObject:
//             folder   -> [name, dateAdded, dateLastAccessed, "YES", rgb-or-"", children] (6)
//             bookmark -> [name, dateAdded, dateLastAccessed, "NO", ref] (5)
//    READ   -parseArray:
//             idx3 via -boolValue selects folder/bookmark;
//             folder reads rgb@4 (""->nil) + children@5; bookmark reads ref@4.
//  The "YES"/"NO" are LITERAL STRINGS (not booleans). Do NOT tidy the asymmetry.
//

import Foundation

@objc(PSBookmarks)
class PSBookmarks: PSBookmarkFolder {

    // MARK: - Singleton

    private static var psDefaultBookmarks: PSBookmarks?

    /// The singleton instance. Obj-C calls `[PSBookmarks defaultBookmarks]`;
    /// Swift (incl. the XCTest guard) calls `PSBookmarks.default()`.
    @objc(defaultBookmarks)
    class func `default`() -> PSBookmarks {
        if psDefaultBookmarks == nil {
            psDefaultBookmarks = PSBookmarks(localBookmarks: ())
        }
        return psDefaultBookmarks!
    }

    // MARK: - Init

    @objc(initLocalBookmarks)
    init(localBookmarks: ()) {
        super.init(name: nil, dateAdded: nil, dateLastAccessed: nil, rgbHexString: nil, children: nil)
        self.name = NSLocalizedString("BookmarksTitle", comment: "")
        let bookmarksPath = (AppPaths.bookmarksPath as NSString).appendingPathComponent("PSBookmarks.plist")
        let dataArray = NSArray(contentsOfFile: bookmarksPath) as? [Any]
        loadBookmarks(fromArray: dataArray)
    }

    @objc(initCloudBookmarks)
    init(cloudBookmarks: ()) {
        super.init(name: nil, dateAdded: nil, dateLastAccessed: nil, rgbHexString: nil, children: nil)
        self.name = NSLocalizedString("BookmarksTitle", comment: "")
        let data = "plistStringToCreateADataThingo".data(using: .utf8)!
        // format should be NSPropertyListXMLFormat_v1_0
        let array = try? PropertyListSerialization.propertyList(from: data, options: [], format: nil)
        loadBookmarks(fromArray: array as? [Any])
    }

    // MARK: - Add / delete

    @objc(addBookmarkObject:withFolderString:)
    class func addBookmarkObject(_ bookmark: PSBookmarkObject, withFolderString folderString: String?) -> Bool {
        var ret = false
        let bookmarks = PSBookmarks.default()
        if folderString == nil || folderString == "" {
            bookmarks.addChild(bookmark)
            ret = true
        } else {
            let parentFolder = PSBookmarks.getBookmarkFolder(forFolderString: folderString)
            parentFolder.addChild(bookmark)
            ret = true
        }
        PSBookmarks.saveBookmarksToFile()
        return ret
    }

    @objc(addBookmarkWithRef:name:folderString:)
    class func addBookmark(withRef r: String?, name n: String?, folderString: String?) -> Bool {
        let date = Date()
        let bookmark = PSBookmark(name: n, dateAdded: date, dateLastAccessed: date, bibleReference: r)
        return PSBookmarks.addBookmarkObject(bookmark, withFolderString: folderString)
    }

    @objc(deleteBookmark:fromFolderString:)
    class func deleteBookmark(_ n: String, fromFolderString folderString: String?) {
        let parent = PSBookmarks.getBookmarkFolder(forFolderString: folderString)
        var array: [Any] = parent.children ?? []
        // need to identify which is the correct child
        for (idx, element) in array.enumerated() {
            if let obj = element as? PSBookmarkObject, obj.name == n {
                array.remove(at: idx)
                break
            }
        }
        parent.children = array
        PSBookmarks.saveBookmarksToFile()
    }

    // MARK: - Lookups

    @objc(getBookmarkFolderForFolderString:)
    class func getBookmarkFolder(forFolderString folderString: String?) -> PSBookmarkFolder {
        let bookmarks = PSBookmarks.default()
        if folderString == nil || folderString == "" || folderString == NSLocalizedString("BookmarksTitle", comment: "") {
            return bookmarks
        }
        let folders = folderString!.components(separatedBy: AppConstants.folderSeparatorString)
        var parentFolder: PSBookmarkFolder = bookmarks
        for folderCount in 0..<folders.count {
            var foundFolder = false
            let kids = parentFolder.children ?? []
            for kidNumber in 0..<kids.count {
                if let kid = kids[kidNumber] as? PSBookmarkObject, kid.name == folders[folderCount] {
                    if let folderKid = kids[kidNumber] as? PSBookmarkFolder {
                        parentFolder = folderKid
                    }
                    foundFolder = true
                    break
                }
            }
            if !foundFolder {
                // if we don't find the next folder, just place the bookmark here. This shouldn't be possible!
                alog("getBookmarkFolderForFolderString: couldn't find the folder - \(folders[folderCount])")
                break
            }
        }
        return parentFolder
    }

    @objc(getHighlightRGBColourStringForBookAndChapterRef:withVerse:)
    class func getHighlightRGBColourString(forBookAndChapterRef bookAndChapterRef: String, withVerse verse: Int) -> String? {
        // Changed to %ld and casting to "long" in order to compensate for 64bit "long" version of NSInteger.
        return PSBookmarks.default().getHighlightRGBColourString(forBookAndChapterRef: bookAndChapterRef,
                                                                 withVerse: String(format: "%ld", verse))
    }

    @objc(getBookmarksForCurrentRef)
    class func getBookmarksForCurrentRef() -> NSMutableArray {
        let currentRef = PSRefHelper.createRefString(PSRefHelper.getCurrentBibleRef())
        return PSBookmarks.getBookmarks(forBookAndChapterRef: currentRef)
    }

    @objc(getBookmarksForBookAndChapterRef:)
    class func getBookmarks(forBookAndChapterRef bookAndChapterRef: String) -> NSMutableArray {
        let bookmarks = PSBookmarks.default()
        return bookmarks.getBookmarks(forBookAndChapterRef: bookAndChapterRef)
    }

    // MARK: - Positional-array (de)serialization (schema-locked)

    @objc(parseArray:)
    func parseArray(_ array: [Any]?) -> PSBookmarkObject? {
        guard let array = array, array.count != 0 else {
            return nil
        }

        let n = array[0] as? String
        let da = array[1] as? Date
        let dla = array[2] as? Date
        let folderString = array[3] as? String
        if (folderString as NSString?)?.boolValue ?? false {
            // tis a folder
            var rgb = array[4] as? String
            if rgb == "" {
                rgb = nil
            }
            let kids = array[5] as? [Any] ?? []
            var kidsArray: [Any] = []
            for child in kids {
                if let kid = parseArray(child as? [Any]) {
                    kidsArray.append(kid)
                }
            }
            return PSBookmarkFolder(name: n, dateAdded: da, dateLastAccessed: dla, rgbHexString: rgb, children: kidsArray)
        } else {
            // tis a bookmark
            let r = array[4] as? String
            return PSBookmark(name: n, dateAdded: da, dateLastAccessed: dla, bibleReference: r)
        }
    }

    private func loadBookmarks(fromArray dataArray: [Any]?) {
        var kidsArray: [Any] = []
        if let dataArray = dataArray {
            for child in dataArray {
                if let kid = parseArray(child as? [Any]) {
                    kidsArray.append(kid)
                }
            }
            self.children = kidsArray
        } else {
            self.children = []
        }
    }

    @objc(parseBookmarkObject:)
    class func parseBookmarkObject(_ bookmarkObject: PSBookmarkObject?) -> NSArray? {
        guard let bookmarkObject = bookmarkObject else {
            return nil
        }
        let capacity = bookmarkObject.folder ? 7 : 5
        let ret = NSMutableArray(capacity: capacity)
        ret.add(bookmarkObject.name as Any)
        ret.add(bookmarkObject.dateAdded as Any)
        ret.add(bookmarkObject.dateLastAccessed as Any)
        if bookmarkObject.folder {
            ret.add("YES")
            if let folder = bookmarkObject as? PSBookmarkFolder, let rgb = folder.rgbHexString {
                ret.add(rgb)
            } else {
                ret.add("")
            }
            let folder = bookmarkObject as? PSBookmarkFolder
            let folderChildren = folder?.children ?? []
            let kids = NSMutableArray(capacity: folderChildren.count)
            for case let child as PSBookmarkObject in folderChildren {
                if let kid = PSBookmarks.parseBookmarkObject(child) {
                    kids.add(kid)
                }
            }
            ret.add(kids)
        } else {
            ret.add("NO")
            if let bookmark = bookmarkObject as? PSBookmark {
                ret.add(bookmark.ref as Any)
            }
        }
        return ret
    }

    // MARK: - lastModified

    class func lastModified(_ bookmarkObject: PSBookmarkObject) -> Date? {
        // Mirrors the Obj-C quirk: compares the recursive childDate but assigns
        // child.dateLastAccessed (not childDate). Preserved verbatim.
        var returnDate = bookmarkObject.dateLastAccessed
        if bookmarkObject.folder, let folder = bookmarkObject as? PSBookmarkFolder {
            for case let child as PSBookmarkObject in (folder.children ?? []) {
                let childDate = PSBookmarks.lastModified(child)
                if let childDate = childDate, let rd = returnDate,
                   childDate.compare(rd) == .orderedDescending {
                    returnDate = child.dateLastAccessed
                }
            }
        }
        return returnDate
    }

    @objc(lastModified)
    class func lastModified() -> Date? {
        let bookmarks = PSBookmarks.default()
        return PSBookmarks.lastModified(bookmarks)
    }

    // MARK: - Persistence

    @objc(saveBookmarksToFile)
    @discardableResult
    class func saveBookmarksToFile() -> Bool {
        let bookmarks = PSBookmarks.default()
        var ret = false
        let bookmarksPath = (AppPaths.bookmarksPath as NSString).appendingPathComponent("PSBookmarks.plist")
        if let children = bookmarks.children, children.count > 0 {
            let data = NSMutableArray(capacity: children.count)
            for case let child as PSBookmarkObject in children {
                if let kid = PSBookmarks.parseBookmarkObject(child) {
                    data.add(kid)
                }
            }
            ret = data.write(toFile: bookmarksPath, atomically: true)
        } else {
            // no bookmarks left! Write out the file to zeros...
            let data = NSMutableArray(capacity: 1)
            ret = data.write(toFile: bookmarksPath, atomically: false)
        }
        dlog("\n-- Bookmarks: finished saveBookmarksToFile")
        return ret
    }

    @objc(importBookmarksFromV2)
    class func importBookmarksFromV2() {
        let oldBookmarks = UserDefaults.standard.array(forKey: "bookmarks2")
        guard let oldBookmarks = oldBookmarks else {
            return
        }
        let bookmarks = PSBookmarks.default()
        var createImportedFolder = true
        for case let obj as PSBookmarkObject in (bookmarks.children ?? []) {
            if obj.name == NSLocalizedString("BookmarksImportedFolderName", comment: "") {
                // if there already exists an @"imported" folder, don't recreate it!
                createImportedFolder = false
                break
            }
        }
        if createImportedFolder {
            let importFolder = PSBookmarkFolder(name: NSLocalizedString("BookmarksImportedFolderName", comment: ""),
                                                dateAdded: Date(),
                                                dateLastAccessed: Date(),
                                                rgbHexString: nil,
                                                children: nil)
            _ = PSBookmarks.addBookmarkObject(importFolder, withFolderString: nil)
        }
        for case let ref as String in oldBookmarks {
            _ = PSBookmarks.addBookmark(withRef: PSRefHelper.createRefString(ref),
                                        name: ref,
                                        folderString: NSLocalizedString("BookmarksImportedFolderName", comment: ""))
        }
        UserDefaults.standard.removeObject(forKey: "bookmarks2")
        UserDefaults.standard.synchronize()
    }
}
